File: src/Test/PolymorphismFFI.php
<?php
$exports['runPolymorphismFFI'] = function($limit) {
    $n = (int)$limit;
    $dict = (object)[
        "mempty_" => 1,
        "mappend_" => function($x) {
            return function($y) use ($x) {
                return $x + $y;
            };
        }
    ];
    $polyLoop = function($dictionary, $count, $acc) {
        while ($count !== 0) {
            $acc = ($dictionary->mappend_)($acc)($dictionary->mempty_);
            $count--;
        }
        return $acc;
    };
    return $polyLoop($dict, $n, 0);
};
return $exports;

File: src/Test/PolymorphismFFI.fs
module Test.PolymorphismFFI

type Monoidish<'a> = { mempty: 'a; mappend: 'a -> 'a -> 'a }
let rec polyLoop dictionary count acc =
    if count = 0 then acc
    else polyLoop dictionary (count - 1) (dictionary.mappend acc dictionary.mempty)
let runPolymorphismFFI (n: obj) =
    polyLoop { mempty = 1; mappend = (+) } (unbox<int> n) 0 :> obj
